nexus/bin/qmca

```diff
@@ -1842,16 +1845,55 @@ class DatAnalyzer(QBase):
             v = data.LocalEnergy_sq - data.LocalEnergy**2
             data.Variance  = v
             stats.Variance = self.stat_value(v[nbe:])
-        #end if            
-        if 'LocalEnergy' in data and 'ElecElec' in data and 'MPC' in data and 'KEcorr' in data:
-            v = data.LocalEnergy - data.ElecElec + data.MPC + data.KEcorr
-            data.CorrectedEnergy  = v
-            stats.CorrectedEnergy = self.stat_value(v[nbe:])
-        elif 'LocalEnergy' in data and 'ElecElec' in data and 'MPC' in data:
-            v = data.LocalEnergy - data.ElecElec + data.MPC
-            data.CorrectedEnergy  = v
-            stats.CorrectedEnergy = self.stat_value(v[nbe:])
         #end if
+
+        # find finite size corrected energy
+        if 'LocalEnergy' in data and ('MPC' in data or 'KEcorr' in data):
+            ce = None
+            if 'Kinetic' in data:
+                non_ee_energies = self.non_ee_potentials
+                nee = data.Kinetic.copy()
+                for pot in self.non_ee_potentials:
+                    if pot in data:
+                        nee += data[pot]
+                    #end if
+                #end for
+                # allow the correction only if the local energy sum is correct
+                le_recovered = False
+                le_ref = data.LocalEnergy.sum()
+                for ee in ('ElecElec','MPC'):
+                    if ee in data:
+                        le_sum = (nee + data[ee]).sum()
+                        le_recovered |= abs((le_sum-le_ref)/le_ref)<1e-8
+                    #end if
+                #end if
+                if le_recovered:
+                    ce = nee
+                    if 'MPC' in data:
+                        ce += data.MPC
+                    elif 'ElecElec' in data:
+                        ce += data.ElecElec
+                    #end if
+                    if 'KEcorr' in data:
+                        ce += data.KEcorr
+                    #end if
+                #end if
+            #end if
+            if ce is None:
+                ce = data.LocalEnergy.copy()
+                if 'MPC' in data and 'ElecElec' in data:
+                    ce += data.MPC - data.ElecElec
+                #end if
+                if 'KEcorr' in data:
```

Comment:
`if 'KEcorr' in data` and the above one can be added at the end of the giant if block once by itself.

Reply:
True.  I wanted to keep the two attempts complete and separate, but I don't care much either way.